File: ais3-pre-exam/2023/crypto/2des/exp/exp.js
const crypto = require('crypto')
let { res, hint, iv, hint_pt } = require('./output')

iv = Buffer.from(iv, 'hex')
res = Buffer.from(res, 'hex')
hint = Buffer.from(hint, 'hex')
hint_pt = Buffer.from(hint_pt, 'hex')

function encrypt(msg, key, iv) {
    const cipher = crypto.createCipheriv('des-cbc', key, iv)
    let encrypted = cipher.update(msg)
    encrypted = Buffer.concat([encrypted, cipher.final()])
    return encrypted
}

function decrypt(msg, key, iv) {
    const decipher = crypto.createDecipheriv('des-cbc', key, iv)
    let decrypted = decipher.update(msg, 'nyan~')
    decrypted = Buffer.concat([decrypted, decipher.final()])
    return decrypted
}

function solve() {
    const middle = new Map()
    console.log('Generating middle map...')
    for (let a = 0b11110000; a < 0b11111111; a+=2)
    {
        process.stdout.write(`a = ${a}\r`)
        for (let b = 0b11110000; b < 0b11111111; b+=2)
        for (let c = 0b11110000; c < 0b11111111; c+=2)
        for (let d = 0b11110000; d < 0b11111111; d+=2)
        for (let f = 0b11110000; f < 0b11111111; f+=2)
        for (let g = 0b11110000; g < 0b11111111; g+=2)
        for (let h = 0b11110000; h < 0b11111111; h+=2)
        for (let i = 0b11110000; i < 0b11111111; i+=2)
        {
            const key1 = Buffer.from([a, b, c, d, f, g, h, i])
            const encrypted = encrypt(hint_pt, key1, iv)
                middle.set(encrypted.toString('hex'), key1)
        }
    }
        
    
    console.log('Brute forcing...')
    for (let a = 0b11110000; a < 0b11111111; a+=2)
    {
        process.stdout.write(`a = ${a}\r`)
        for (let b = 0b11110000; b < 0b11111111; b+=2)
        for (let c = 0b11110000; c < 0b11111111; c+=2)
        for (let d = 0b11110000; d < 0b11111111; d+=2)
        for (let f = 0b11110000; f < 0b11111111; f+=2)
        for (let g = 0b11110000; g < 0b11111111; g+=2)
        for (let h = 0b11110000; h < 0b11111111; h+=2)
        for (let i = 0b11110000; i < 0b11111111; i+=2)
        {
            const key2 = Buffer.from([a, b, c, d, f, g, h, i])
            try {
                const decrypted = decrypt(hint, key2, iv).toString('hex')
                if (middle.has(decrypted)) {
                    const key1 = middle.get(decrypted)
                    console.log(key1.toString('hex'))
                    const flag = decrypt(decrypt(res, key2, iv), key1, iv)
                    console.log(flag.toString('utf8'))
                    return
                }
            } catch (e) {
                // console.log(e)
            }
        }
    }
}

solve()

File: ais3-pre-exam/2023/crypto/2des/exp/exp.py
from Crypto.Cipher import DES
from Crypto.Util.Padding import pad, unpad

res = '6020e9735ca3bf2f63aebcf3622c994880ffed2b509c91414c75d4c500ee80f4'
hint_pt = '414953337b3f3f3f3f3f3f3f3f3f3f7d'
hint = '118cd68957ac93b269335416afda70e6d79ad65a09b0c0c6c50917e0cee18c93'
iv = '4149533320e4b889'

res = bytes.fromhex(res)
hint_pt = bytes.fromhex(hint_pt)
hint = bytes.fromhex(hint)
iv = bytes.fromhex(iv)


def encrypt(key, iv, plaintext) -> bytes:
    cipher = DES.new(key, DES.MODE_CBC, iv)
    ciphertext = cipher.encrypt(pad(plaintext, DES.block_size))
    return ciphertext


def decrypt(key, iv, ciphertext):
    cipher = DES.new(key, DES.MODE_CBC, iv)
    plaintext = unpad(cipher.decrypt(ciphertext), DES.block_size)
    return plaintext


def main():
    middle = {}
    for i in range(0b11110000, 0b11111111, 2):
        print(f'{i}/255')
        for j in range(0b11110000, 0b11111111, 2):
            for k in range(0b11110000, 0b11111111, 2):
                for l in range(0b11110000, 0b11111111, 2):
                    for m in range(0b11110000, 0b11111111, 2):
                        for n in range(0b11110000, 0b11111111, 2):
                            for o in range(0b11110000, 0b11111111, 2):
                                for p in range(0b11110000, 0b11111111, 2):
                                    key1 = bytes([i, j, k, l, m, n, o, p])
                                    middle[encrypt(key1, iv, hint_pt)] = key1

    for i in range(0b11110000, 0b11111111, 2):
        print(f'{i}/255')
        for j in range(0b11110000, 0b11111111, 2):
            for k in range(0b11110000, 0b11111111, 2):
                for l in range(0b11110000, 0b11111111, 2):
                    for m in range(0b11110000, 0b11111111, 2):
                        for n in range(0b11110000, 0b11111111, 2):
                            for o in range(0b11110000, 0b11111111, 2):
                                for p in range(0b11110000, 0b11111111, 2):
                                    key2 = bytes([i, j, k, l, m, n, o, p])
                                    try:
                                        if decrypt(key2, iv, hint) in middle:
                                            key1 = middle[decrypt(
                                                key2, iv, hint)]
                                            print(f"key1: {key1.hex()}")
                                            print(f"key2: {key2.hex()}")
                                            m = decrypt(
                                                key1, iv, decrypt(key2, iv, res))
                                            print(f"flag: {m.decode()}")
                                            return
                                    except:
                                        continue


if __name__ == '__main__':
    main()
